### ISSUE TITLE
`Sized` not detected for fixed size array size

### ISSUE THREAD

**Issue by aticu** [rust-lang/rust#61427]
The `Sized` bound does not seem to work correctly when using `std::mem::size_of` to determine the size of a fixed size array.

The following code fails:

```Rust
fn breaks<T: Sized>() {
    let _: [u8; std::mem::size_of::<T>()];
}
```

with the message

```text
error[E0277]: the size for values of type `T` cannot be known at compilation time
 --> src/main.rs:2:17
  |
2 |     let _: [u8; std::mem::size_of::<T>()];
  |                 ^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: the trait `std::marker::Sized` is not implemented for `T`
  = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
  = help: consider adding a `where T: std::marker::Sized` bound
  = note: required by `std::mem::size_of`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
```

while this code works:

```Rust
fn works<T: Sized>() {
    let _ = std::mem::size_of::<T>();
}
```

I'd expect the Sized bound to work the same in both cases. Also even if that was not possible, the error message is not particularly helpful in this case, since it contains false information (`T: Sized` is clearly true).

## Meta

Tested with the latest nightly and stable.

### Nightly

```text
rustc 1.37.0-nightly (7840a0b75 2019-05-31)
binary: rustc
commit-hash: 7840a0b753a065a41999f1fb6028f67d33e3fdd5
commit-date: 2019-05-31
host: x86_64-unknown-linux-gnu
release: 1.37.0-nightly
LLVM version: 8.0
```

### Stable

```
rustc 1.35.0 (3c235d560 2019-05-20)
binary: rustc
commit-hash: 3c235d5600393dfe6c36eeed34042efad8d4f26e
commit-date: 2019-05-20
host: x86_64-unknown-linux-gnu
release: 1.35.0
LLVM version: 8.0
```

**Comment by varkor** [comment]
This is https://github.com/rust-lang/rust/issues/43408.

**Comment by aticu** [comment]
I didn't find that in my search. Sorry for the duplicate.

I will close this then.
